src/service/position-observer-impl.js

```diff
@@ -0,0 +1,410 @@
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element === element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      if (this.entries_[i].element === element) {
+        this.entries_[i].handler = null;
+        this.entries_.splice(i, 1);
+        if (this.entries_.length == 0) {
+          this.stopCallback();
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot unobserve unobserved element');
+  }
+
+  /** @param {!Object} unusedEntry */
+  updateEntryPosition_(unusedEntry) {}
+
+  startCallback() {}
+
+  stopCallback() {}
+
+  increaseFidelityCallback() {}
+
+  decreaseFidelityCallback() {}
+}
+
+export class AmpDocPositionObserver extends AbstractPositionObserver {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+
+    super(ampdoc);
+
+    /** @private {boolean} */
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {number} */
+    this.scrollTimer_ = Date.now();
+
+    /** @private {Array<function()>} */
+    this.unlisteners_ = [];
+  }
+
+  /* @override */
+  startCallback() {
+    // listen to viewport scroll event to help pass determine if need to
+    const stopScroll = () => {
+      if (Date.now() - this.scrollTimer_ < 500) {
+        // viewport scroll in the last 500 ms, wait
+        return;
+      }
+      // assume scroll stops, if no scroll event in the last 500,
+      this.inScroll_ = false;
+    };
+
+    this.unlisteners_.push(this.viewport_.onScroll(() => {
+      this.inScroll_ = true;
+      this.scrollTimer_ = Date.now();
+      this.schedulePass_();
+      // TODO(@zhouyx, #9208) using debounce.
+      setTimeout(stopScroll.bind(this), 500);
+    }));
+    this.unlisteners_.push(this.viewport_.onChanged(() => {
+      this.vsync_.measure(() => {
+        this.pass_(true);
+      });
+    }));
+    // QQQ: Already call this.updateEntryPosition_(entry) in observe()? Needed?
+    // initial pass
+    this.vsync_.measure(() => {
+      this.pass_(true);
+    });
+  }
+
+  /* @override */
+  stopCallback() {
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /** @param {boolean=} recursive */
+  schedulePass_(recursive) {
+    // TODO (@zhouyx, #9208):
+    // P1: account for effective fidelity using this.effectiveFidelity
+    // P2: do passes on onDomMutation (if available using MutationObserver)
```

Comment:
This is ok for in-a-box host but probably not very useful for AMP, since most of changes happen due to CSS and most of DOM mutations are constraint to the AMP elements. Instead, we need to tie this to the future Layers "invalidate DOM area" concept.